M06 async

mwand opened the pull request:
# Asynchronous Programming: we've been teaching it all wrong (again)

After many false starts, I finally realized that trying to explain async programming in terms of promises is like explaining typescript in terms of assembly language: it's just the **wrong level**

Among other things, it fails to emphasize that you can determine the points at which your execution can be interrupted is **statically determinable**

These slides start by explaining what the uninterruptible segments of your program are.

We take "run to completion" as a fundamental concept (explaining that we mean "run to the completion of the critical section"), and go on from there.

In this pull request, we have just the slides and the scripts-- no repo or anything else.  You'll see that at some point the slides are unchanged; I want to get you all's feedback before putting in any more work.

I'm very excited by this.

See if you think this could work!!

jon-bell commented:
Very nice! A few comments:

Slide 4: Nice to see this same graphic that I made in Keynote 10 years ago is still relevant :) If interested, can add a new technical development: call it 20,000ns for reading data from NVME

I think that the part that you are iterating on is the part that students seem to struggle less with, however, as opposed to: async races. For example, there are plenty of ways in React to end up with a trace of:
- Page loads
- Requests data
- Renders
- Data received (Page not correctly updated to reflect updated data)

How do students avoid these? How do they debug these?

abhutta0 commented:
TODO: Add examples in module page.

abhutta0 commented:
I have added animations in few slides. More comments:
I still think that slide 30 should be moved after slide 13 (and before promise.all one)
S14: There might be a mistake here. First two "time passes" shouldn't be there. 3 request received should be printed together without delay. May be you can test the code to verify.
I think we should add the slides back for the activity, if we are using the same activity as last semester. I didn't find them in "Old Extra" section.

abhutta0 commented:
S30/31 either version is fine.
S14: comments might be misleading then. I think last version had some console log that said "request made". The idea is that these get printed when we "call" then asyn fn, not when we "return" from them (or finish them). Whatever we can do to make that distinction.

mwand commented:
Note I've separated the code for the activity from the code for the examples.

mwand commented:
> S8: needs notes

added

> S9: fonts sizes are different in text that points to green boxes

fixed

> S12: It says: "..pause point corresponds to the end of ..." how about adding something like: "run to completion" means that you to the end of this before you can pause.

I think that's what I said.
> S23 and S26 are essentially same. We we need S26

Kept first version (S23)

> S28 can go after slide 30 (before put it all together)

I think the current flow is fine

> Old Extra section can go

Done

abhutta0 commented:
feel free to merge to main
